File: src/rolling_apply_specialized.c
/* 
 * This implementation is a subset of the implementation described in
 * "Algorithms for Unevenly-Spaced Time Series", Eckner (2011).
 */

#include <math.h>
#include <stdlib.h>

#ifndef swap
#define swap(a,b) {temp=(a); (a)=(b); (b)=temp;}
#endif


/******************* Helper functions ********************/


// Return smallest element of an array
double array_min(double values[], int n)
{
  // values ... array of values
  // n      ... length of array
  
  if (n == 0)
    return INFINITY;
  
  double min_value = values[0];    
  for (int i = 1; i < n; i++) {
    if (values[i] < min_value)
      min_value = values[i];
  }
  return min_value;
}


/*
Find the k-th largest (counting starts at zero) of an array using the "quickselect" algorithm
-) O(N) average case performance
-) the input array will be rearranged
*/
double quickselect(double values[], int n, int k)
{
  // values ... array of values
  // n      ... length of array
  // k      ... return k-th smallest element
  
  int i, j, left, right, mid;
  double pivot, temp;
  left = 0;
  right = n - 1;
  
  while (1 > 0) {
    if (right <= left + 1) {
      // Array down to 1-2 elements
      if ((right == left + 1) && (values[right] < values[left]))
        swap(values[left], values[right])
      return values[k];
    } else {
      // Select pivot element
      mid = (left + right) / 2;   // integer devision
      swap(values[mid], values[left + 1]);
      
      // Put pivot element and the elements at the left and right boundary in the correct relative order
      if (values[left] > values[right])
        swap(values[left], values[right])
      if (values[left + 1] > values[right])
        swap(values[left + 1], values[right])
      if (values[left] > values[left + 1])
        swap(values[left], values[left + 1])
      
      // Put smaller elements to left of pivot, larger to right
      i = left + 1;
      j = right;
      pivot = values[left + 1];
      for (;;) {
        do i++; while (values[i] < pivot);
        do j--; while (values[j] > pivot);
        if (j < i)
          break;
        swap(values[i], values[j])
      }
      values[left + 1] = values[j];
      values[j] = pivot;
      if (j >= k)
        right = j-1;
      if (j <= k)
        left = i;
    }
  }
}


// Find the median value of an array (which gets scrambled)
double median(double values[], int n)
{
  // values ... array of values
  // n      ... length of array

  double value_low, value_high;
  
  // Determine the mid points of the array
  int mid_low = (n - 1) / 2;
  int mid_high = n - mid_low - 1;
  value_low = quickselect(values, n, mid_low);
  
  if (mid_low < mid_high) {   // even number of elements -> two mid points
    // Get the smallest element to the right of lowest mid-point
    value_high = array_min(values + mid_high, n - mid_high);
    return (value_low + value_high) / 2; 
  } else
    return value_low;
}


/****************** END: Helper functions ****************/


// Rolling number of observation values
void rolling_num_obs(double values[], double times[], int *n, double values_new[],
  double *width_before, double *width_after)
{
  // values       ... array of time series values
  // times        ... array of observation times
  // n            ... number of observations, i.e. length of 'values' and 'times'
  // values_new   ... array of length *n to store output time series values
  // width_before ... (non-negative) width of rolling window before t_i
  // width_after  ... (non-negative) width of rolling window after t_i
  
  int left = 0, right = -1;
  
  for (int i = 0; i < *n; i++) {   
    // Shrink window on the left
    while ((left < *n) && (times[left] <= times[i] - *width_before))
      left++;
    
    // Expand window on the right
    while ((right < *n - 1) && (times[right + 1] <= times[i] + *width_after))
      right++;
    
    // Number of observations is equal to length of window
    values_new[i] = right - left + 1;
  }
}


// Rolling sum of observation values
void rolling_sum(double values[], double times[], int *n, double values_new[],
  double *width_before, double *width_after)
{
  // values       ... array of time series values
  // times        ... array of observation times
  // n            ... number of observations, i.e. length of 'values' and 'times'
  // values_new   ... array of length *n to store output time series values
  // width_before ... (non-negative) width of rolling window before t_i
  // width_after  ... (non-negative) width of rolling window after t_i
  
  int left = 0, right = -1;
  double roll_sum = 0;
  
  for (int i = 0; i < *n; i++) {
    // Expand window on the right
    while ((right < *n - 1) && (times[right + 1] <= times[i] + *width_after)) {
      right++;
      roll_sum = roll_sum + values[right];
    }
    
    // Shrink window on the left
    while ((left < *n) && (times[left] <= times[i] - *width_before)) {
      roll_sum = roll_sum - values[left];
      left++;
    }
    
    // Update rolling sum
    values_new[i] = roll_sum;
  }
}


// Rolling average of observation values
void rolling_mean(double values[], double times[], int *n, double values_new[],
  double *width_before, double *width_after)
{
  // values       ... array of time series values
  // times        ... array of observation times
  // n            ... number of observations, i.e. length of 'values' and 'times'
  // values_new   ... array of length *n to store output time series values
  // width_before ... (non-negative) width of rolling window before t_i
  // width_after  ... (non-negative) width of rolling window after t_i
  
  int left = 0, right = -1;
  double roll_sum = 0;
  
  for (int i = 0; i < *n; i++) {
    // Expand window on the right
    while ((right < *n - 1) && (times[right + 1] <= times[i] + *width_after)) {
      right++;
      roll_sum = roll_sum + values[right];
    }
    
    // Shrink window on the left to get half-open interval
    while ((left < *n) && (times[left] <= times[i] - *width_before)) {
      roll_sum = roll_sum - values[left];
      left++;
    }
    
    // Calculate mean of values in rolling window
    values_new[i] = roll_sum / (right - left + 1);
  }
}


// Rolling maximum of observation values
void rolling_max(double values[], double times[], int *n, double values_new[],
  double *width_before, double *width_after)
{
  // values       ... array of time series values
  // times        ... array of observation times matching time series values
  // n            ... length of 'values'
  // values_new   ... array (of same length as 'values') used to store output
  // width_before ... (non-negative) width of rolling window before t_i
  // width_after  ... (non-negative) width of rolling window after t_i
  
  int j, left = 0, right = -1, max_pos = 0;
  
  for (int i = 0; i < *n; i++) {
    // Expand window on the right
    while ((right < *n - 1) && (times[right + 1] <= times[i] + *width_after)) {
      right++;
      if (values[right] >= values[max_pos])
        max_pos = right;
    }
    
    // Shrink window on the left to get half-open interval
    while ((left < *n) && (times[left] <= times[i] - *width_before))
      left++;      
    
    // Recalculate position of maximum if old maximum dropped out
    // Inline functionality of max_index() to avoid function call overhead
    if (max_pos < left) {
      max_pos = left;
      for (j = left+1; j <= right; j++)
        if (values[j] >= values[max_pos])
          max_pos = j;
    }
    
    // Save maximum for current time window
    if (left <= right)  // non-empty window
      values_new[i] = values[max_pos];
    else                // empty window
      values_new[i] = -INFINITY;
  }
}


// Rolling minimum of observation values
void rolling_min(double values[], double times[], int *n, double values_new[],
  double *width_before, double *width_after)
{
  // values       ... array of time series values
  // times        ... array of observation times matching time series values
  // n            ... length of 'values'
  // values_new   ... array (of same length as 'values') used to store output
  // width_before ... (non-negative) width of rolling window before t_i
  // width_after  ... (non-negative) width of rolling window after t_i
  
  int j, left = 0, right = -1, min_pos = 0;
  
  for (int i = 0; i < *n; i++) {   
    // Expand window on the right
    while ((right < *n - 1) && (times[right + 1] <= times[i] + *width_after)) {
      right++;
      if (values[right] <= values[min_pos])
        min_pos = right;
    }
    
    // Shrink window on the left to get half-open interval
    while ((left < *n) && (times[left] <= times[i] - *width_before))
      left++;      
    
    // Recalculate position of minimum if old minimum dropped out
    // Inline the calculation of the minimum position to avoid any function call overhead
    if (min_pos < left) {
      min_pos = left;
      for (j = left+1; j <= right; j++)
        if (values[j] <= values[min_pos])
          min_pos = j;
    }
    
    // Save minium for current time window
    if (left <= right)  // non-empty window
      values_new[i] = values[min_pos];
    else                // empty window
      values_new[i] = INFINITY;
  }
}


// Rolling median
void rolling_median(double values[], double times[], int *n, double values_new[], 
  double *width_before, double *width_after)
{
  // values       ... array of time series values
  // times        ... array of observation times matching time series values
  // n            ... length of 'values'
  // values_new   ... array (of same length as 'values') used to store output
  // width_before ... (non-negative) width of rolling window before t_i
  // width_after  ... (non-negative) width of rolling window after t_i
  
  int j, window_length, left = 0, right = -1;
  double values_tmp[*n];      // temporary array for median(), which shuffles the input data 

  for (int i = 0; i < *n; i++) {
    // Expand window on the right
    while ((right < *n - 1) && (times[right + 1] <= times[i] + *width_after))
      right++;
    
    // Shrink window on the left end
    while ((left < *n) && (times[left] <= times[i] - *width_before))
      left++;
    
    // Copy data in rolling window to temporary array, and calculate the median
    window_length = right - left + 1;
    for (j = 0; j < window_length; j++)
      values_tmp[j] = values[left + j];
    values_new[i] = median(values_tmp, window_length);
  }
}

File: R/rolling_apply_specialized.R
#############################################################################
# Specialized implementations of rolling_apply() for certain choices of FUN #
#############################################################################

#' Apply Rolling Function (Specialized Implementation)
#' 
#' This function provides a fast, specialized implementation of \code{\link{rolling_apply}} for certain choices of \code{FUN} and for \code{by=NULL} (i.e. when moving the rolling time window one observation at a time, rather than by a fixed temporal amount).
#' 
#' It is usually not necessary to call this function, because it is called automatically by \code{\link{rolling_apply}} whenever a specialized implementation is available.
#' 
#' @param x a numeric time series object with finite, non-NA observation values.
#' @param width a finite, positive \code{\link[lubridate]{duration}} object, specifying the temporal width of the rolling time window.
#' @param FUN a function to be applied to the vector of observation values inside the half-open (open on the left, closed on the right) rolling time window.
#' @param align either \code{"right"}, \code{"left"}, or \code{"center"}. Specifies the alignment of each output time relative to its corresponding time window. Using \code{"right"} gives a causal (i.e. backward-looking) time series operator, while using \code{"left"} gives a purely forward-looking time series operator.
#' @param interior logical. Should time windows lie entirely in the interior of the temporal support of \code{x}, i.e. inside the time interval \code{[start(x), end(x)]}?
#' @param \ldots further arguments passed to or from methods.
#' 
#' @references Eckner, A. (2010) \emph{Algorithms for Unevenly Spaced Time Series: Moving Averages and Other Rolling Operators}. 
#' @keywords internal
rolling_apply_specialized <- function(x, ...) UseMethod("rolling_apply_specialized")


#' @describeIn rolling_apply_specialized Implementation for \code{"uts"} objects with finite, non-NA observation values.
#' 
#' @examples
#' rolling_apply_specialized(ex_uts(), dhours(12), FUN=length)
#' rolling_apply_specialized(ex_uts(), dhours(12), FUN=length, align="center")
#' rolling_apply_specialized(ex_uts(), dhours(12), FUN=length, align="left")
#' 
#' rolling_apply_specialized(ex_uts(), dhours(12), FUN=length)
#' rolling_apply_specialized(ex_uts(), dhours(12), FUN=length, interior=TRUE)
#' 
#' # Rolling sum
#' rolling_apply_specialized(ex_uts(), ddays(1), FUN=sum)
#' rolling_apply_specialized(ex_uts(), ddays(1), FUN=sum) - rolling_apply(ex_uts(), ddays(1), FUN=sum)
#' 
#' # Rolling min/max
#' rolling_apply_specialized(ex_uts(), ddays(1), FUN=min)
#' rolling_apply_specialized(ex_uts(), ddays(1), FUN=max)
rolling_apply_specialized.uts <- function(x, width, FUN, align="right", interior=FALSE, ...)
{
  # Extract the name of the function to be called
  if (is.function(FUN)) {
    if (identical(FUN, length))
      FUN <- "length"
    else if (identical(FUN, mean))
      FUN <- "mean"
    else if (identical(FUN, min))
      FUN <- "min"
    else if (identical(FUN, max))
      FUN <- "max"
    else if (identical(FUN, median))
      FUN <- "median"
    else if (identical(FUN, sum))
      FUN <- "sum"
    else {
      FUN <- deparse(substitute(FUN))
      if (length(FUN) > 1)
        stop("Custom functions (FUN) are not supported")
    }
  }
  
  # Select C function
  if (FUN == "length")
    C_fct <- "rolling_num_obs"
  else if (FUN == "min")
    C_fct <- "rolling_min"
  else if (FUN == "max")
    C_fct <- "rolling_max"
  else if (FUN == "mean")
    C_fct <- "rolling_mean"
  else if (FUN == "median")
    C_fct <- "rolling_median"
  else if (FUN == "sum")
    C_fct <- "rolling_sum"
  else
    stop("This function does not have a specialized rolling_apply() implementation")
  
  # Determine the window width before and after the current output time, depending on the window alignment
  check_window_width(width)
  if (align == "right") {
    width_before <- width
    width_after <- 0
  } else if (align == "left") {
    width_before <- 0
    width_after <- width
  } else if (align == "center") {
    width_before <- width / 2
    width_after <- width / 2
  } else
    stop("'align' has to be either 'left', 'right', or 'center")
    
  
  # Call C function
  out <- generic_C_interface(x, width_before=width_before, width_after=width_after, C_fct=C_fct)
  
  # Replace NaN by NA in output to be consistent with generic rolling_apply()
  out$values[is.nan(out$values)] <- NA
  
  # Optionally, drop output times for which the corresponding time window is not completely inside the temporal support of x
  if (interior)
    out <- window(out, start=start(out) + width_before, end(out) - width_after)
  out
}


#' Specialized Rolling Apply Available?
#' 
#' Check whether \code{\link{rolling_apply_specialized.uts}} can be called for a given \code{\link{uts}} object with arguments \code{FUN} and \code{by}.
#' 
#' @param x a \code{"uts"} object.
#' @param FUN see \code{\link{rolling_apply_specialized}}.
#' @param by see \code{\link{rolling_apply_specialized}}.
#' 
#' @keywords internal
#' @examples 
#' have_rolling_apply_specialized(ex_uts(), FUN=mean)
#' have_rolling_apply_specialized(ex_uts(), FUN="mean")
#' have_rolling_apply_specialized(ex_uts(), FUN=mean, by=ddays(1))
#' have_rolling_apply_specialized(uts(NA, Sys.time()), FUN=mean)
#' 
#' FUN <- mean
#' have_rolling_apply_specialized(ex_uts(), FUN=FUN)
have_rolling_apply_specialized <- function(x, FUN, by=NULL)
{
  # Extract the name of the function to be called
  if (is.function(FUN)) {
    if (identical(FUN, length))
      FUN <- "length"
    else if (identical(FUN, mean))
      FUN <- "mean"
    else if (identical(FUN, min))
      FUN <- "min"
    else if (identical(FUN, max))
      FUN <- "max"
    else if (identical(FUN, median))
      FUN <- "median"
    else if (identical(FUN, sum))
      FUN <- "sum"
    else
      FUN <- deparse(substitute(FUN))
  }
  
  # Determine if fast special purpose implementation is available
  (length(FUN) == 1) && (FUN %in% c("length", "mean", "min", "max", "median", "sum")) &&
    is.null(by) && (is.numeric(x$values)) && (!anyNA(x$values)) && (all(is.finite(x$values)))
}

